Fix segfault on using constants for variable parameters in various situations

GiudGiud opened the pull request:
…n, closes #22597

GiudGiud committed:
Fix segfault on calling nonAD version of thermal expansion eigenstrain, closes #22597

hugary1995 commented:
Yep, right fix.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1224454/) on ed16180 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22598/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1224467/) on ed16180 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8ebf4</th>
      <th colspan=4>#22598 ed1618</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.36%</td>
      <td>84.36%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80674</td>
      <td>80678</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14956</td>
      <td>14952</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22598/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22598/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8ebf4</th>
      <th colspan=4>#22598 ed1618</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.59%</td>
      <td>89.59%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26417</td>
      <td>26420</td>
      <td>+3</td>
      <td>8</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3069</td>
      <td>3069</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22598/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22598/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22598/coverage/xfem)

</details>



This comment will be updated on new commits.

dschwen commented:
Oh, because of constant coupling... (should be required coupled anyways)

dschwen commented on modules/tensor_mechanics/src/materials/ComputeThermalExpansionEigenstrainBase.C:
```diff
@@ -34,9 +34,10 @@ ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::ComputeThermalExpansionEigen
-    _deigenstrain_dT(is_ad ? nullptr
-                           : &this->template declarePropertyDerivative<RankTwoTensor>(
-                                 _eigenstrain_name, this->getVar("temperature", 0)->name())),
+    _deigenstrain_dT((is_ad || !(this->getVar("temperature", 0)))
+                         ? nullptr
+                         : &this->template declarePropertyDerivative<RankTwoTensor>(
+                               _eigenstrain_name, this->getVar("temperature", 0)->name())),
```
we should have a `isCoupledConstant` helper method in `Coupleable` to make such code more readable.

hugary1995 replied:
Yeah, we had a discussion in the referenced issue. Why don't you join the party there?

GiudGiud commented:
Making it required wont stop people from passing constants. 

You prefer is coupledConstant over a coupledName() that doesnt seg fault?
At the end of the day we re trying to get devs to avoid getVar()->name() so it needs to be the most convenient

dschwen commented:
> You prefer is coupledConstant over a coupledName() that doesnt seg fault?
No, we need both! What is a `coupledName` "that doesn't segfault" supposed to do? We still need some sort of check to deterimine if the derivative property should be created or not.

GiudGiud commented:
I feel like it s pretty likely code downstream will expect the derivative to have been created, but that is caught nicer than a seg fault so I m fine with that. 

I agree there's a use for both routines. 
I ll amend this PR and the issue to reflect the broader problem

dschwen commented:
> I feel like it s pretty likely code downstream will expect the derivative to have been created

No, it is not. Derivatives that do not exist are assumed to be zero. That's why this works at all with constant passed in in your fix.

dschwen commented:
By the way, we cannot currently mix constants and variables names in vector coupling (multiple coupled components). We could capture this with a custom `CoupledVariableName` class if we add another field for coupled constant value.

dschwen commented:
That would effectively be a `std::variant` ...

GiudGiud committed:
Add coupledName interface to get variable names
Add iscoupledconstant interface to check whether a user passed in a constant instead of a variable
Closes #22597

GiudGiud committed:
Remove getVar()->name() from framework source as it segfaults on constant, refs #22597

GiudGiud committed:
Remove getVar()->name() from modules source as it segfaults on constant, refs #22597

GiudGiud commented:
I added the new APIs and propagated the coupledName() in moose and the modules. 

While tensor mechanics is good about using 0s for non-existent material properties, looking at the other modules affected, it s clear some expect the material properties to exist. So I did not add checks on isCoupledConstant everywhere, so some code will be declaring properties like dC_d800. This is not really an issue, it does not change the results or segfault like previously. 

I did not take care of the mix of constant and variables. This is a long standing request by others but outside the scope of this work (which is removing those seg faults with constants)

GiudGiud commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -2140,10 +2140,30 @@ Coupleable::coupledIndices(const std::string & var_name) const
 Coupleable::coupledNames(const std::string & var_name) const
 {
-  auto func = [this, &var_name](unsigned int comp) { return getVar(var_name, comp)->name(); };
+  auto func = [this, &var_name](unsigned int comp)
```
very true. Updated

dschwen commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -2140,10 +2140,22 @@ Coupleable::coupledIndices(const std::string & var_name) const
+{
+  if (getVar(var_name, comp))
+    return getVar(var_name, comp)->name();
+  // For constants, we ll just return the constant as the name
```
```suggestion
  // For constants, we'll just return the constant as the name
```

dschwen commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -2140,10 +2140,22 @@ Coupleable::coupledIndices(const std::string & var_name) const
+  else if (MooseUtils::parsesToReal(var_name))
+    return var_name;
+  else
+    mooseError("Variable '", var_name, "' does not exist, yet its coupled name is requested");
```
You have access to the MooseObject pointer to the derived class that issues the failing call. Outputting its name could be helpful for the user.

dschwen commented on framework/src/materials/PiecewiseLinearInterpolationMaterial.C:
```diff
@@ -47,7 +47,7 @@ PiecewiseLinearInterpolationMaterial::PiecewiseLinearInterpolationMaterial(
     _extrap(getParam<bool>("extrapolation")),
     _property(declareProperty<Real>(_prop_name)),
-    _dproperty(declarePropertyDerivative<Real>(_prop_name, getVar("variable", 0)->name()))
+    _dproperty(declarePropertyDerivative<Real>(_prop_name, coupledName("variable", 0)))
```
So for constant coupling this would declare something like `dprop/d45.7`?

dschwen replied:
the pattern should probalbly be 
```suggestion
    _dproperty(isCoupledConstant("variable") ? nullptr : &declarePropertyDerivative<Real>(_prop_name, coupledName("variable", 0)))
```

and changing `_dproperty` to a pointer

dschwen commented on modules/heat_conduction/src/materials/SemiconductorLinearConductivity.C:
```diff
@@ -36,7 +36,7 @@ SemiconductorLinearConductivity::SemiconductorLinearConductivity(const InputPara
     _electric_conductivity(declareProperty<Real>(_base_name + "electrical_conductivity")),
     _delectric_conductivity_dT(declarePropertyDerivative<Real>(
-        _base_name + "electrical_conductivity", getVar("temp", 0)->name()))
+        _base_name + "electrical_conductivity", coupledName("temp", 0)))
```
same here

dschwen replied:
and many other places

GiudGiud replied:
changed everywhere. Causes Marmot to break on one file. 
Can you make the patch or should I revert on file here?

GiudGiud replied:
*everywhere
Only on the declare*Derivatives. The get*Derivative should be fine querying dCd47.7

GiudGiud committed:
Avoid declare material property derivatives if the derivatee (new word) is constant (in MODULES)

GiudGiud committed:
Avoid declaring material property derivatives if the variable we take the derivative with is actually a constant (in framework/)

GiudGiud committed:
Address review:
- fix comment
- mention object name in errors for coupleable
except for:
- tag-related routines which are more geared towards devlpt at this point
- not-implemented errors which are also only hit by devs who dont need to know which object

GiudGiud commented:
ah I dont think I have access to Marmot but it should be a really easy fix, _dDdc is now a pointer inherited by PoreGBMrtx..Material

dschwen commented:
Working on it now.

dschwen commented:
Ok, I think I have the appropriate fix: (censored for export control reasons

```
-  MaterialProperty<RealTensorValue> & _dMdc;
+  MaterialProperty<RealTensorValue> * _dMdc;
```

```
-    _dMdc(declarePropertyDerivative<RealTensorValue>(_M_name, _c_name)),
+    _dMdc(isCoupledConstant(_c_name)
+              ? nullptr
+              : &declarePropertyDerivative<RealTensorValue>(_M_name, _c_name)),
```

```
-    _dMdc[_qp] = ########### * _dDdc[_qp];
+    if (_dMdc)
+      (*_dMdc)[_qp] = ########### * (*_dDdc)[_qp];
```

dschwen commented:
Ok, I can update the MOOSE submodule on my marmot branch as soon as this get's merged.

GiudGiud commented:
sounds good. So patch ready?

we need to break next
Once this is approved i ll seek permission

dschwen commented:
patch ready

dschwen commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -2140,10 +2140,23 @@ Coupleable::coupledIndices(const std::string & var_name) const
+  if (getVar(var_name, comp))
+    return getVar(var_name, comp)->name();
+  // For constants, we'll just return the constant as the name
+  else if (MooseUtils::parsesToReal(var_name))
+    return var_name;
+  else
+    mooseError(
+        _c_name, ": Variable '", var_name, "' does not exist, yet its coupled name is requested");
+}
```
I wonder if we should make `coupledName`  error out if a constant was coupled. I.e. remove line 2149 and 2150. This would force users to use `isCoupledConstant`. 

We could potentially add `coupledConstant` to recover the coupled constant value.

I just don;t see a usecase for working with a string representation of a coupled constant number literal.

dschwen replied:
I can see new users wondering why material properties like `dF/d12.721` are created.

dschwen replied:
You could keep the lines but make 2150 emit a different `mooseError`. E.g. "The name of a coupled variable was requested, but the user provided a coupled constant value instead."

GiudGiud replied:
This is more general than material properties though. coupledName can be used in unrelated code. Do we want to ban constants as soon as the routine relies on the name (for an error message for example?) of the variable?

GiudGiud replied:
do dF/d12.711 appear from the getMaterialDerivative calls? Do they get created by default instead of a 0 matprop?

dschwen replied:
you would have to suppress creation with a call to isCoupledConstant. Erroring would force the dev to do that. Also, are we using variable names in error messages? We could have a separate call for that which returns the constant string. In fact we already do, it is: `getParam<std::vector<VariableName>>("var_param_name")[component]`

GiudGiud replied:
There are 100 or so occurences of getVar->name(), which because they seg fault with constants have been changed to coupledName().

If we made them error instead, it's  definitely better than a seg fault, but it's lost capability. We could no longer use constants there for a while until the devs fix them all. 

Which is unlikely to happen, because the devs did not fix it when it was a seg fault, so when it s a benign error..

imo, the next improvement should not lie with coupledName() but with declareMatPropDerivative, that would avoid declaring a derivative if the variable is a constant.

dschwen replied:
How is this lost capable of it never worked?!

GiudGiud replied:
It works now :)

GiudGiud replied:
> Erroring would force the dev to do that.

The seg fault did not stop the devs from doing it.

I'll summarize my position on this:
- I want coupledName to be used because the getVar path is not working
- I dont want the spread of coupledName to limit our ability to pass constants as variables

GiudGiud replied:
@dschwen I caved in
Can I please get an approval :)

dschwen replied:
I agree with both of your bullets, but the spread of coupledName is not limiting our ability to pass constants as variables! In the cases where we used getVar instead of coupledName we didn't ever have that capability in the first place! We'd get a segfault instead of an error. Prior to this PR to have this capability the developer would have to explicitly check for a nullptr, after this PR the developer will have to use `isCoupledConstant`.

Overall the PR is an improvement because you are
1. turning a segfault into a proper mooseError
2. more readable functions will be used that convey intent

dschwen commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -2140,10 +2140,23 @@ Coupleable::coupledIndices(const std::string & var_name) const
 Coupleable::coupledNames(const std::string & var_name) const
 {
-  auto func = [this, &var_name](unsigned int comp) { return getVar(var_name, comp)->name(); };
+  auto func = [this, &var_name](unsigned int comp) { return coupledName(var_name, comp); };
```
The behavior described above would not be an issue here, as we currently do not allow mixing of variable and constant coupling in one coupled variable vector.

GiudGiud replied:
this code would work even when we allow the mixing if we do not disallow coupledName on a constant

GiudGiud commented:
Let's go?

GiudGiud committed:
Error if the name of a constant passed in lieu of a variable is queried

GiudGiud committed:
Add a test for trying to get a coupled name on a missing variable / a constant

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1300383/) on 75fe5d6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22598/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22598/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cb680b795f4783b35c51eb60d902cb432855abe3`
